proposals/p5337.md

```diff
@@ -0,0 +1,967 @@
+the interface or type argument must be in the same file. For example:
+
+```carbon
+class BigInt { ... }
+
+interface IntLike {
+  // Only valid in the same library as `ImplicitAs`
+  // or `BigInt`.
+  extend impl as ImplicitAs(BigInt);
+}
+```
+
+The
+[additional restrictions on final impls](/docs/design/generics/details.md#libraries-that-can-contain-a-final-impl)
+mean that `extend final impl as` can only be used in the same library as the
+interface being extended, not the type arguments.
+
+Similarly, the expression to the right of the `as` must correspond to a single
+interface, due to
+[leads issue #4566](https://github.com/carbon-language/carbon-lang/issues/4566)
+and [proposal #5168](https://github.com/carbon-language/carbon-lang/pull/5168).
+
+### Name conflicts
+
+If two associated constants or functions have the same name in the two
+interfaces, this is an error since there is no way to assign values to both in
+an `impl` of the extending interface.
+
+Example with an associated function:
+
+```carbon
+interface A {
+  fn F();
+}
+
+interface B {
+  fn F();
+  // Name conflict with `F`.
+  extend impl as A;
+}
+
+class C {
+  impl as B {
+    // Probably refers to the `F` explicitly declared in `B` --
+    // no way to implement the `F` that comes from `A`.
+    fn F() { ... }
+  }
+}
+```
+
+Example with an associated constant:
+
+```carbon
+interface A {
+  let T:! type;
+}
+
+interface B {
+  let T:! type;
+  // Name conflict with `T`.
+  extend impl as A;
+}
+
+class C {
+  impl as B {
+    // Two different associated constants named `T` and no way
+    // to distinguish them.
+    where T = i32;
+  }
+}
+```
+
+Notice that the rewrite to put the members of `A` into `B` would result in a
+conflict. Further, if there is only one member with that name after the rewrite
+in `B`, then there is no way to write the generated `impl` to set the member of
+`A` from the values of the associated constants in `B`.
+
+However, if the extending interface gives a value to the associated constant,
+there is no need to specify that value in an `impl`, so that is not an error.
+
+```carbon
+interface A {
+  let T:! type;
+  fn F() -> T;
+}
+
+interface B {
+  let T:! type;
+  // Okay
+  extend impl as A where .T = T*;
+}
+
+class C {
+  impl as B {
+    // `T` here is `B.T`
+    where T = i32;
+    // `F` here is `B.F` and `A.F`.
+    fn F() -> i32*;
+  }
+}
+```
+
+This is equivalent to:
+
+```carbon
+interface A {
+  let T:! type;
+  fn F() -> T;
+}
+
+interface B {
+  let T:! type;
+  // Skip A.T due to conflict.
+  fn F() -> T*;
+}
+
+impl forall [U:! B] U as A where .T = .(B.T)* {
+  fn F() -> T = U.(B.F);
+}
+
+class C {
+  impl as B {
+    where T = i32;
+    fn F() -> i32*;
+  }
+}
+```
+
+This was used in the `OutputForwardIterator` example in
+[the "Proposal" section](#proposal).
+
+Name conflicts are not expected to be a problem in practice since this form of
+extension requires the two interfaces to be defined in the same file.
```

Comment:
Only for `extend final impl`, right?

Reply:
Added some qualifiers. If we do think this is going to be a problem, though, I'm not sure what to suggest. If we are copying the members automatically we potentially have a problem where names will conflict in the face of evolution. If we are not copying the members automatically, we are doing it manually, which seems even more of a problem since any change will cause it to fall out of date.